Convert rct_optimizations to be a ROS-generic CMake package

schornakj opened the pull request:
The first step at enabling #35. The work I had done previously converted the RCT packages to be ROS2 packages via Ament, which isn't an ideal strategy for us to use here, so this is a fresh take on the task.

- Remove Catkin dependencies, macros, and functions from `rct_optimizations`.
- Change minimum required CMake version from 2.8.3 to 3.5.0.
- Add a set of CMake macros (in `rct_optimizations/cmake/rct_macros.cmake`) to streamline package configuration and test setup. The idea is that the other RCT packages will be able to use these too. These macros are lightly modified from the ones used in [Tesseract](https://github.com/ros-industrial-consortium/tesseract/blob/master/tesseract/tesseract_common/cmake/tesseract_macros.cmake).
- Use a more "modern CMake" approach in `rct_optimizations/CMakeLists.txt`. It now provides its libraries through the `rct::rct_optimizations` target.
- Move test source code to a `test` directory, which is added as a subdirectory if the `RCT_ENABLE_RUN_TESTING` flag is `True`.
- Change how `rct_image_tools`, `rct_ros_tools`, and `rct_examples` find and link against `rct_optimizations` to accommodate the fact that it is no longer a ROS package.

TODOs:

- [x] Questions for @marip8: Do the changes I made to the way tests are structured still allow CI to build and run tests? Do we have to add `RCT_ENABLE_RUN_TESTING` to `.github_actions.yml`?

marip8 commented on rct_optimizations/CMakeLists.txt:
```diff
@@ -42,48 +22,32 @@ add_library(${PROJECT_NAME}
-  FILES_MATCHING PATTERN "*.h"
+install(DIRECTORY include/${PROJECT_NAME}
+  DESTINATION include
+  FILES_MATCHING PATTERN "*.h" PATTERN "*.hpp"
   PATTERN ".svn" EXCLUDE
```
I would remove these lines. There is currently nothing in this directory we don't want to install and we're not using SVN

marip8 commented on rct_optimizations/test/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+#find_package(GTest REQUIRED)
```
Remove commented code

marip8 commented on rct_optimizations/CMakeLists.txt:
```diff
@@ -42,48 +22,32 @@ add_library(${PROJECT_NAME}
-
-  catkin_add_gtest(${PROJECT_NAME}_extrinsic_camera_on_wrist_tests test/extrinsic_camera_on_wrist_utest.cpp)
-  target_link_libraries(${PROJECT_NAME}_extrinsic_camera_on_wrist_tests ${PROJECT_NAME}_test_support)
+if(RCT_ENABLE_RUN_TESTING)
+  enable_testing()
+  rct_add_run_tests_target()
+  add_subdirectory(test)
 endif()
```
I would also consider separating the running of the unit tests from the building of the tests. I'm finding right now that I want to build the tests often and run individual tests offline. This can be problematic because but they take 2-3 minutes in total to run. I don't want to wait that long every time I build unit tests just to see ones that I don't care about at the time run

schornakj replied:
The macros do support separately building and running tests, I just need to change where they're included in relation to the `RCT_ENABLE_RUN_TESTING` flag.

One challenge is that I have to make sure that GTest can be found and linked against whenever we build the tests, which is a little tricky because its authors have conceptually designed it to need to be downloaded and built from source, vs. providing precompiled binaries.

schornakj replied:
This should be addressed in my latest changes.

schornakj commented:
I've made some changes which I will push up once I get tests running through CI again:
- I added a `rct_common` package, which now contains CMake macros and the GTest external package.
- I changed the CMake arguments for building and running tests. Setting `RCT_BUILD_TESTS=True` will build the tests, and setting `RCT_RUN_TESTS=True` will cause them to be run using `make test` after they're built.

schornakj commented:
CI currently fails with the error below, which I think is because the environment used in the Github Action doesn't have `git` installed.

```
Errors     << rct_common:cmake /root/target_ws/logs/rct_common/build.cmake.000.log
  CMake Error at /usr/share/cmake-3.5/Modules/ExternalProject.cmake:1757 (message):
  
error: could not find git for clone of GTest
```

marip8 commented:
That seems odd. Doesn't the Docker use git to clone the repository its building as well as packages from .rosinstall files?

You can specify that additional packages be installed before running the build using the `ADDITIONAL_DEBS` environment variable with `industrial_ci`. Maybe try that with `git`?

schornakj commented:
You're right, I didn't understand it quite correctly and git is already installed at the top level.

schornakj commented:
@marip8 I rebased onto the latest master branch and fixed the issue with linking to GTest on the xenial job.

I think that the `rct_optimizations` tests are being, with two caveats:
- Setting `RCT_RUN_TESTS=True` makes the tests run both when building and during the `run_tests` step, which makes the CI process take longer.
- The `run_tests` step doesn't appear to report that the `rct_optimizations` tests ran and succeeded.
